### ISSUE TITLE
Use an OpenFOAM dictionary for configuration

### ISSUE THREAD

**Issue by MakisH** [precice/openfoam-adapter#30]
We currently use [yaml-cpp](https://github.com/jbeder/yaml-cpp) to parse our `precice-adapter-config.yml` YAML file (see more in our [wiki](https://github.com/precice/openfoam-adapter/wiki/Configuration))

Although YAML is a nice language to use for configuration files, there are some problems:

* yaml-cpp is an additional, not so common dependency
* the yaml-cpp version may conflict with the Boost version (see again this [issue](https://github.com/precice/openfoam-adapter/issues/18#issuecomment-384728248))
* it is a format foreign to OpenFOAM (although this could also be seen as an advantage)

Therefore, it would be good to allow configuring the adapter through a dictionary. This could be done alongside or instead our current YAML-based file.

**Comment by MakisH** [comment]
In this context, we should also include an additional option for the kind of each couplingDataUser.

Currently, we use the name (e.g. `Forces0`) to extract the kind (`Force`).

Thanks to @DavidSCN for reporting a relevant problem.

**Comment by MakisH** [comment]
A few weeks back I made some progress on this. The latest work-in-progress state can be found in the branch [config-dict](https://github.com/precice/openfoam-adapter/tree/config-dict).

I am currently stuck in transforming the nested lists logic of YAML to dictionaries. For example:
```yaml
interfaces:
    - mesh: Solid-to-Inner-Fluid
      patch: inner-interface
      read-data: [Sink-Temperature-Inner-Fluid, Heat-Transfer-Coefficient-Inner-Fluid]
      write-data: [Sink-Temperature-Solid, Heat-Transfer-Coefficient-Solid]
    - mesh: Solid-to-Outer-Fluid
      patch: outer-interface
      read-data: [Sink-Temperature-Outer-Fluid, Heat-Transfer-Coefficient-Outer-Fluid]
      write-data: [Sink-Temperature-Solid, Heat-Transfer-Coefficient-Solid]
```
This example is from the [heat exchanger tutorial](https://github.com/precice/tutorials/blob/master/CHT/heat_exchanger/buoyantSimpleFoam-CalculiX/config.yml) (for CalculiX, but the logic is the same).

What I originally wanted to do was represent this as a list of dictionaries. This could be quite tricky and a better idea (@olesenm) would be to have these dictionaries unconnected to each other and then get them from the table of contents. But I still have have not tried to implement this.

**Comment by olesenm** [comment]
To clarify, a list of dictionaries isn't intrinsically more difficult. You can either read them as an entry (see https://develop.openfoam.com/Development/OpenFOAM-plus/blob/master/src/OpenFOAM/db/dictionary/dictionaryListEntry/dictionaryListEntry.H) and them loop over them. Another way is to have the same syntax and use that entry to create a PtrList of your interfacing objects. Finally can just have them as a collection of subdicts (as discussed).

One advantage of the collection of subdicts is that you will able to use changeDictionary to rewrite/modify the entries. If they are a list (with the '(...)' delimiters) they are locked away from easy dictionary manipulation.
On the other side, if you handle them as a PtrList, the meaning of the keywords are more easily handled by the interfacing object (construct from dictionary). 

Probably worth the effort to combine both. Have subdicts, but pass off them each off to a construct from dictionary and save as a PtrList.

**Comment by MakisH** [comment]
This was closed with #105.
